api/api-environments-v2/src/main/java/com/thoughtworks/go/apiv2/environments/EnvironmentsControllerV2.java

```diff
@@ -0,0 +1,234 @@
+                currentUsername(),
+                operationResult
+        );
+
+        EnvironmentConfig updateConfigElement = fetchEntityFromConfig(environmentName);
+
+        setEtagHeader(updateConfigElement, response);
+
+        return handleCreateOrUpdateResponse(request, response, updateConfigElement, operationResult);
+    }
+
+    public String remove(Request request, Response response) {
+        String environmentName = request.params("name");
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+
+        EnvironmentConfig  environmentConfig = fetchEntityFromConfig(environmentName);
+
+        environmentConfigService.deleteEnvironment(environmentConfig, currentUsername(),
+                    result);
+        return handleSimpleMessageResponse(response, result);
+    }
+
+    @Override
+    public String etagFor(EnvironmentConfig entityFromServer) {
+        return entityHashingService.md5ForEntity(entityFromServer);
+    }
+
+    @Override
+    public EnvironmentConfig doFetchEntityFromConfig(String name) {
+        ConfigElementForEdit<EnvironmentConfig> mergedEnvironmentforDisplay = environmentConfigService.getMergedEnvironmentforDisplay(name, new HttpLocalizedOperationResult());
+
+        return mergedEnvironmentforDisplay == null ? null : mergedEnvironmentforDisplay.getConfigElement();
+    }
+
+    @Override
+    public EnvironmentConfig buildEntityFromRequestBody(Request req) {
+        JsonReader jsonReader = GsonTransformer.getInstance().jsonReaderFrom(req.body());
+        return EnvironmentRepresenter.fromJSON(jsonReader);
+    }
+
+    @Override
+    public Consumer<OutputWriter> jsonWriter(EnvironmentConfig environmentConfig) {
+        return writer -> EnvironmentRepresenter.toJSON(writer, environmentConfig);
+    }
+
+    private void haltIfEntityWithSameNameExists(EnvironmentConfig environmentConfig) {
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+        ConfigElementForEdit<EnvironmentConfig> existingEnvironmentConfig = environmentConfigService.getMergedEnvironmentforDisplay(environmentConfig.name().toString(), result);
+
+        if (existingEnvironmentConfig == null) {
+            return;
+        }
+
+        environmentConfig.addError("name", format("Environment name should be unique. Environment with name '%s' already exists.", environmentConfig.name().toString()));
+        throw haltBecauseEntityAlreadyExists(jsonWriter(environmentConfig), "environment", environmentConfig.name().toString());
+    }
+
+    private String calculateEtag(List<EnvironmentConfig> environmentConfigs) {
+        final String environmentConfigSegment = environmentConfigs
+                .stream()
+                .map(this::etagFor)
+                .collect(Collectors.joining(SEP_CHAR));
+
+        return DigestUtils.sha256Hex(StringUtils.joinWith(SEP_CHAR, currentUsername().getUsername(), environmentConfigSegment));
```

Comment:
@vishaldevgire @kritika-singh3 -- This ETag Computation has a problem.

Environments can be viewed only by super admins. And the view of the environments doesn't change based on the super administrator accessing it (unlike dashboard).
Adding the user information while computing ETag will result into different ETags based on the user who is accessing it.

IMO, the ETag computation logic for environments should be user agnostic. And should result in same ETag.

Reply:
Corrected.